git/revlist.go

```diff
@@ -117,3 +120,207 @@ func revListCallback(c *Client, opt RevListOptions, commits []CommitID, excludeL
 	}
 	return nil
 }
+
+// Internal pre-computed record of which symbolic references
+//  are pointing to a particular ref name (without the refs/ prefix)
+var symRefsForRefName map[string]string
+
+// Internal pre-computed record of which ref names are pointing to
+//  to a particular commit. Values are in the form of "HEAD -> master, branch2,
+//  tag: tag1"
+var refNamesForCommit map[string]string
+
+// We have been asked to pre-compute the ref names for important
+//  commits as a way to improve performance.
+func buildRefNames(c *Client) error {
+	symRefsForRefName = make(map[string]string)
+	refNamesForCommit = make(map[string]string)
+
+	// TODO support more symbolic refs than just HEAD
+	headrefspec, err := SymbolicRefGet(c, SymbolicRefOptions{Short: true}, "HEAD")
+	if err != nil {
+		return err
+	}
+	symRefsForRefName[string(headrefspec)] = "HEAD"
+
+	branches, err := c.GetBranches()
+	if err != nil {
+		return err
+	}
+
+	for _, branch := range branches {
+		branchName := string(branch)
+
+		commitId, err := RefSpec(branchName).Value(c)
```

Comment:
We should be more careful about "RefSpec" vs "Ref" (although there's still a lot of places in older code that use the word "RefSpec" instead of "Ref")..a Ref is a single reference in a repo. A RefSpec is (or at least should be) a spec that can include wildcards, but is often used incorrectly in this repo from before I understood the difference.

Reply:
Yes, I think I now see the difference. A Ref is a pointer file in the .git/refs directory. A RefSpec is something that may or may not match Ref. I think that RefSpec can include both source and destinations as I can see from your PR and after reading more in the git manuals.

Reply:
Yeah, there's a lot of places in the code that need that cleaned up, because I didn't really understand that at first either. At some point it might be worth just ripping out everything other than "Src" and "Dst" from the Refspec type and then fixing all the compile errors will force us to be more careful.

(Also, refs can be packed with newer versions of git, so they don't necessarily live under .git/refs, but conceptually.. yeah.)